crates/turborepo-discovery/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+//! turborepo-discovery
+//!
+//! This package contains a number of strategies for discovering various things
+//! about a workspace. These traits come with a basic implementation and some
+//! adaptors that can be used to compose them together.
+//!
+//! This powers various intents such as 'query the daemon for this data, or
+//! fallback to local discovery if the daemon is not available'. Eventually,
+//! these strategies will implement some sort of monad-style composition so that
+//! we can track areas of run that are performing sub-optimally.
+
+#![feature(async_fn_in_trait)]
+#![feature(return_position_impl_trait_in_trait)]
+
+use tokio_stream::{iter, StreamExt};
+use turbopath::AbsoluteSystemPathBuf;
+use turborepo_repository::package_manager::{self, PackageManager};
+
+#[derive(Clone)]
+pub struct PackageData {
+    pub package_json: AbsoluteSystemPathBuf,
+    pub turbo_json: Option<AbsoluteSystemPathBuf>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("discovery unavailable")]
+    Unavailable,
+    #[error("discovery failed")]
+    Failed,
```

Comment:
Planning on hashing out a better plan for this. These larger components failing are probably something we want to track somehow. Will speak to @NicholasLYang about plans for this.